File: ClientSDK/JS/tunrtc-client.js
/**
 * TunRTC JavaScript SDK
 * Open-source WebRTC client for voice and video communication
 * @version 1.0.0
 */

class TunRTCClient {
    constructor(config) {
        this.apiUrl = config.apiUrl || 'https://localhost:7000/api';
        this.hubUrl = config.hubUrl || 'https://localhost:7000/hubs/signaling';
        this.token = null;
        this.connection = null;
        this.localStream = null;
        this.peerConnections = new Map();
        this.iceServers = [];
        this.sessionId = null;
        this.userId = null;
        this.username = null;
        
        // Event callbacks
        this.onLocalStream = null;
        this.onRemoteStream = null;
        this.onParticipantJoined = null;
        this.onParticipantLeft = null;
        this.onParticipantStatusChanged = null;
        this.onMessage = null;
        this.onError = null;
        this.onConnectionStateChanged = null;
    }

    /**
     * Register a new user
     */
    async register(username, email, password) {
        try {
            const response = await fetch(`${this.apiUrl}/auth/register`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ username, email, password })
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.message || 'Registration failed');
            }

            const data = await response.json();
            this.token = data.token;
            this.username = data.username;
            return data;
        } catch (error) {
            this._handleError('Register failed', error);
            throw error;
        }
    }

    /**
     * Login with credentials
     */
    async login(email, password) {
        try {
            const response = await fetch(`${this.apiUrl}/auth/login`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ email, password })
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.message || 'Login failed');
            }

            const data = await response.json();
            this.token = data.token;
            this.username = data.username;
            return data;
        } catch (error) {
            this._handleError('Login failed', error);
            throw error;
        }
    }

    /**
     * Get ICE servers configuration
     */
    async getIceServers() {
        try {
            const response = await fetch(`${this.apiUrl}/iceservers`, {
                headers: { 'Authorization': `Bearer ${this.token}` }
            });

            if (!response.ok) {
                throw new Error('Failed to get ICE servers');
            }

            const config = await response.json();
            this.iceServers = config.iceServers;
            return this.iceServers;
        } catch (error) {
            this._handleError('Get ICE servers failed', error);
            throw error;
        }
    }

    /**
     * Create a new session
     */
    async createSession(name, maxParticipants = 10, type = 1) {
        try {
            const response = await fetch(`${this.apiUrl}/session/create`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.token}`
                },
                body: JSON.stringify({ name, maxParticipants, type })
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.message || 'Create session failed');
            }

            const session = await response.json();
            return session;
        } catch (error) {
            this._handleError('Create session failed', error);
            throw error;
        }
    }

    /**
     * Join an existing session
     */
    async joinSession(sessionId, audioEnabled = true, videoEnabled = true) {
        try {
            this.sessionId = sessionId;

            // Get ICE servers
            await this.getIceServers();

            // Get local media stream
            this.localStream = await navigator.mediaDevices.getUserMedia({
                audio: audioEnabled,
                video: videoEnabled
            });

            if (this.onLocalStream) {
                this.onLocalStream(this.localStream);
            }

            // Connect to SignalR hub
            await this._connectToHub();

            // Join session via API
            const response = await fetch(`${this.apiUrl}/session/join`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.token}`
                },
                body: JSON.stringify({ sessionId })
            });

            if (!response.ok) {
                throw new Error('Join session failed');
            }

            const session = await response.json();

            // Join SignalR group
            await this.connection.invoke('JoinSession', sessionId);

            return session;
        } catch (error) {
            this._handleError('Join session failed', error);
            throw error;
        }
    }

    /**
     * Leave current session
     */
    async leaveSession() {
        try {
            if (!this.sessionId) return;

            // Stop local stream
            if (this.localStream) {
                this.localStream.getTracks().forEach(track => track.stop());
                this.localStream = null;
            }

            // Close all peer connections
            this.peerConnections.forEach(pc => pc.close());
            this.peerConnections.clear();

            // Leave SignalR group
            if (this.connection) {
                await this.connection.invoke('LeaveSession', this.sessionId);
            }

            // Leave via API
            await fetch(`${this.apiUrl}/session/leave`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.token}`
                },
                body: JSON.stringify({ sessionId: this.sessionId })
            });

            this.sessionId = null;
        } catch (error) {
            this._handleError('Leave session failed', error);
            throw error;
        }
    }

    /**
     * Update participant status
     */
    async updateStatus(isMuted, isVideoEnabled) {
        try {
            if (!this.sessionId || !this.connection) return;

            // Update local tracks
            if (this.localStream) {
                this.localStream.getAudioTracks().forEach(track => {
                    track.enabled = !isMuted;
                });
                this.localStream.getVideoTracks().forEach(track => {
                    track.enabled = isVideoEnabled;
                });
            }

            // Notify others
            await this.connection.invoke('UpdateStatus', this.sessionId, isMuted, isVideoEnabled);
        } catch (error) {
            this._handleError('Update status failed', error);
            throw error;
        }
    }

    /**
     * Send a chat message
     */
    async sendMessage(message) {
        try {
            if (!this.sessionId || !this.connection) return;

            await this.connection.invoke('SendMessage', this.sessionId, message);
        } catch (error) {
            this._handleError('Send message failed', error);
            throw error;
        }
    }

    /**
     * Connect to SignalR hub
     */
    async _connectToHub() {
        this.connection = new signalR.HubConnectionBuilder()
            .withUrl(this.hubUrl, {
                accessTokenFactory: () => this.token
            })
            .withAutomaticReconnect()
            .build();

        // Handle participant joined
        this.connection.on('ParticipantJoined', async (data) => {
            console.log('Participant joined:', data);
            
            if (this.onParticipantJoined) {
                this.onParticipantJoined(data);
            }

            // Create peer connection and send offer
            await this._createPeerConnection(data.connectionId);
            await this._createOffer(data.connectionId);
        });

        // Handle participant left
        this.connection.on('ParticipantLeft', (data) => {
            console.log('Participant left:', data);
            
            if (this.onParticipantLeft) {
                this.onParticipantLeft(data);
            }

            // Close peer connection
            const pc = this.peerConnections.get(data.connectionId);
            if (pc) {
                pc.close();
                this.peerConnections.delete(data.connectionId);
            }
        });

        // Handle WebRTC offer
        this.connection.on('ReceiveOffer', async (data) => {
            console.log('Received offer from:', data.fromConnectionId);
            await this._handleOffer(data.fromConnectionId, data.offer);
        });

        // Handle WebRTC answer
        this.connection.on('ReceiveAnswer', async (data) => {
            console.log('Received answer from:', data.fromConnectionId);
            await this._handleAnswer(data.fromConnectionId, data.answer);
        });

        // Handle ICE candidate
        this.connection.on('ReceiveIceCandidate', async (data) => {
            await this._handleIceCandidate(data.fromConnectionId, data.candidate);
        });

        // Handle participant status change
        this.connection.on('ParticipantStatusChanged', (data) => {
            if (this.onParticipantStatusChanged) {
                this.onParticipantStatusChanged(data);
            }
        });

        // Handle chat messages
        this.connection.on('ReceiveMessage', (data) => {
            if (this.onMessage) {
                this.onMessage(data);
            }
        });

        // Handle errors
        this.connection.on('Error', (data) => {
            this._handleError('SignalR error', data);
        });

        // Connection state changed
        this.connection.onreconnecting(() => {
            if (this.onConnectionStateChanged) {
                this.onConnectionStateChanged('reconnecting');
            }
        });

        this.connection.onreconnected(() => {
            if (this.onConnectionStateChanged) {
                this.onConnectionStateChanged('connected');
            }
        });

        this.connection.onclose(() => {
            if (this.onConnectionStateChanged) {
                this.onConnectionStateChanged('disconnected');
            }
        });

        await this.connection.start();
        console.log('Connected to SignalR hub');
    }

    /**
     * Create peer connection
     */
    async _createPeerConnection(connectionId) {
        const pc = new RTCPeerConnection({ iceServers: this.iceServers });

        // Add local stream
        if (this.localStream) {
            this.localStream.getTracks().forEach(track => {
                pc.addTrack(track, this.localStream);
            });
        }

        // Handle ICE candidates
        pc.onicecandidate = (event) => {
            if (event.candidate) {
                this.connection.invoke('SendIceCandidate', connectionId, event.candidate);
            }
        };

        // Handle remote stream
        pc.ontrack = (event) => {
            console.log('Received remote track from:', connectionId);
            if (this.onRemoteStream) {
                this.onRemoteStream({
                    connectionId,
                    stream: event.streams[0]
                });
            }
        };

        // Handle connection state
        pc.onconnectionstatechange = () => {
            console.log(`Connection state with ${connectionId}:`, pc.connectionState);
        };

        this.peerConnections.set(connectionId, pc);
        return pc;
    }

    /**
     * Create and send offer
     */
    async _createOffer(connectionId) {
        const pc = this.peerConnections.get(connectionId);
        if (!pc) return;

        const offer = await pc.createOffer();
        await pc.setLocalDescription(offer);
        await this.connection.invoke('SendOffer', connectionId, offer);
    }

    /**
     * Handle incoming offer
     */
    async _handleOffer(connectionId, offer) {
        let pc = this.peerConnections.get(connectionId);
        if (!pc) {
            pc = await this._createPeerConnection(connectionId);
        }

        await pc.setRemoteDescription(new RTCSessionDescription(offer));
        const answer = await pc.createAnswer();
        await pc.setLocalDescription(answer);
        await this.connection.invoke('SendAnswer', connectionId, answer);
    }

    /**
     * Handle incoming answer
     */
    async _handleAnswer(connectionId, answer) {
        const pc = this.peerConnections.get(connectionId);
        if (!pc) return;

        await pc.setRemoteDescription(new RTCSessionDescription(answer));
    }

    /**
     * Handle incoming ICE candidate
     */
    async _handleIceCandidate(connectionId, candidate) {
        const pc = this.peerConnections.get(connectionId);
        if (!pc) return;

        await pc.addIceCandidate(new RTCIceCandidate(candidate));
    }

    /**
     * Handle errors
     */
    _handleError(message, error) {
        console.error(message, error);
        if (this.onError) {
            this.onError({ message, error });
        }
    }

    /**
     * Disconnect and cleanup
     */
    async disconnect() {
        await this.leaveSession();
        
        if (this.connection) {
            await this.connection.stop();
            this.connection = null;
        }

        this.token = null;
        this.username = null;
    }
}

// Export for use in Node.js or browser
if (typeof module !== 'undefined' && module.exports) {
    module.exports = TunRTCClient;
}

File: ClientSDK/CSharp/TunRTCClient.cs
using Microsoft.AspNetCore.SignalR.Client;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using TunRTC.Client.Models;

namespace TunRTC.Client;

public class TunRTCClient : IAsyncDisposable
{
    private readonly HttpClient _httpClient;
    private readonly string _apiUrl;
    private readonly string _hubUrl;
    private HubConnection? _hubConnection;
    private string? _token;
    private string? _sessionId;

    // Events
    public event EventHandler<ParticipantEvent>? ParticipantJoined;
    public event EventHandler<ParticipantEvent>? ParticipantLeft;
    public event EventHandler<ParticipantStatusEvent>? ParticipantStatusChanged;
    public event EventHandler<SignalingMessage>? OfferReceived;
    public event EventHandler<SignalingMessage>? AnswerReceived;
    public event EventHandler<SignalingMessage>? IceCandidateReceived;
    public event EventHandler<ChatMessage>? MessageReceived;
    public event EventHandler<Exception>? ErrorOccurred;
    public event EventHandler<string>? ConnectionStateChanged;

    public bool IsConnected => _hubConnection?.State == HubConnectionState.Connected;
    public string? SessionId => _sessionId;
    public string? Token => _token;

    public TunRTCClient(string apiUrl, string hubUrl)
    {
        _apiUrl = apiUrl.TrimEnd('/');
        _hubUrl = hubUrl;
        _httpClient = new HttpClient
        {
            BaseAddress = new Uri(_apiUrl)
        };
    }

    #region Authentication

    public async Task<AuthResponse> RegisterAsync(string username, string email, string password)
    {
        var request = new RegisterRequest
        {
            Username = username,
            Email = email,
            Password = password
        };

        var response = await _httpClient.PostAsJsonAsync("/auth/register", request);
        response.EnsureSuccessStatusCode();

        var authResponse = await response.Content.ReadFromJsonAsync<AuthResponse>()
            ?? throw new Exception("Failed to deserialize auth response");

        _token = authResponse.Token;
        SetAuthHeader();

        return authResponse;
    }

    public async Task<AuthResponse> LoginAsync(string email, string password)
    {
        var request = new LoginRequest
        {
            Email = email,
            Password = password
        };

        var response = await _httpClient.PostAsJsonAsync("/auth/login", request);
        response.EnsureSuccessStatusCode();

        var authResponse = await response.Content.ReadFromJsonAsync<AuthResponse>()
            ?? throw new Exception("Failed to deserialize auth response");

        _token = authResponse.Token;
        SetAuthHeader();

        return authResponse;
    }

    private void SetAuthHeader()
    {
        if (!string.IsNullOrEmpty(_token))
        {
            _httpClient.DefaultRequestHeaders.Authorization =
                new AuthenticationHeaderValue("Bearer", _token);
        }
    }

    #endregion

    #region Session Management

    public async Task<SessionResponse> CreateSessionAsync(string name, int maxParticipants = 10, SessionType type = SessionType.VideoCall)
    {
        EnsureAuthenticated();

        var request = new CreateSessionRequest
        {
            Name = name,
            MaxParticipants = maxParticipants,
            Type = type
        };

        var response = await _httpClient.PostAsJsonAsync("/session/create", request);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadFromJsonAsync<SessionResponse>()
            ?? throw new Exception("Failed to deserialize session response");
    }

    public async Task<SessionDetailResponse> JoinSessionAsync(string sessionId)
    {
        EnsureAuthenticated();

        _sessionId = sessionId;

        // Join via API
        var request = new JoinSessionRequest { SessionId = sessionId };
        var response = await _httpClient.PostAsJsonAsync("/session/join", request);
        response.EnsureSuccessStatusCode();

        var session = await response.Content.ReadFromJsonAsync<SessionDetailResponse>()
            ?? throw new Exception("Failed to deserialize session response");

        // Connect to SignalR if not already connected
        if (_hubConnection == null || _hubConnection.State == HubConnectionState.Disconnected)
        {
            await ConnectToHubAsync();
        }

        // Join SignalR group
        await _hubConnection!.InvokeAsync("JoinSession", sessionId);

        return session;
    }

    public async Task<SessionDetailResponse> GetSessionAsync(string sessionId)
    {
        EnsureAuthenticated();

        var response = await _httpClient.GetAsync($"/session/{sessionId}");
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadFromJsonAsync<SessionDetailResponse>()
            ?? throw new Exception("Failed to deserialize session response");
    }

    public async Task<List<SessionResponse>> GetActiveSessionsAsync()
    {
        EnsureAuthenticated();

        var response = await _httpClient.GetAsync("/session/active");
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadFromJsonAsync<List<SessionResponse>>()
            ?? throw new Exception("Failed to deserialize sessions response");
    }

    public async Task LeaveSessionAsync()
    {
        if (string.IsNullOrEmpty(_sessionId)) return;

        EnsureAuthenticated();

        // Leave SignalR group
        if (_hubConnection?.State == HubConnectionState.Connected)
        {
            await _hubConnection.InvokeAsync("LeaveSession", _sessionId);
        }

        // Leave via API
        var request = new JoinSessionRequest { SessionId = _sessionId };
        await _httpClient.PostAsJsonAsync("/session/leave", request);

        _sessionId = null;
    }

    public async Task EndSessionAsync(string sessionId)
    {
        EnsureAuthenticated();

        var request = new JoinSessionRequest { SessionId = sessionId };
        var response = await _httpClient.PostAsJsonAsync("/session/end", request);
        response.EnsureSuccessStatusCode();

        _sessionId = null;
    }

    #endregion

    #region ICE Servers

    public async Task<WebRTCConfig> GetIceServersAsync()
    {
        EnsureAuthenticated();

        var response = await _httpClient.GetAsync("/iceservers");
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadFromJsonAsync<WebRTCConfig>()
            ?? throw new Exception("Failed to deserialize ICE servers response");
    }

    #endregion

    #region SignalR Hub

    public async Task ConnectToHubAsync()
    {
        EnsureAuthenticated();

        _hubConnection = new HubConnectionBuilder()
            .WithUrl(_hubUrl, options =>
            {
                options.AccessTokenProvider = () => Task.FromResult(_token);
            })
            .WithAutomaticReconnect()
            .Build();

        RegisterHubHandlers();

        await _hubConnection.StartAsync();
        ConnectionStateChanged?.Invoke(this, "connected");
    }

    private void RegisterHubHandlers()
    {
        if (_hubConnection == null) return;

        _hubConnection.On<ParticipantEvent>("ParticipantJoined", (data) =>
        {
            ParticipantJoined?.Invoke(this, data);
        });

        _hubConnection.On<ParticipantEvent>("ParticipantLeft", (data) =>
        {
            ParticipantLeft?.Invoke(this, data);
        });

        _hubConnection.On<ParticipantStatusEvent>("ParticipantStatusChanged", (data) =>
        {
            ParticipantStatusChanged?.Invoke(this, data);
        });

        _hubConnection.On<SignalingMessage>("ReceiveOffer", (data) =>
        {
            OfferReceived?.Invoke(this, data);
        });

        _hubConnection.On<SignalingMessage>("ReceiveAnswer", (data) =>
        {
            AnswerReceived?.Invoke(this, data);
        });

        _hubConnection.On<SignalingMessage>("ReceiveIceCandidate", (data) =>
        {
            IceCandidateReceived?.Invoke(this, data);
        });

        _hubConnection.On<ChatMessage>("ReceiveMessage", (data) =>
        {
            MessageReceived?.Invoke(this, data);
        });

        _hubConnection.On<object>("Error", (data) =>
        {
            ErrorOccurred?.Invoke(this, new Exception(data.ToString()));
        });

        _hubConnection.Reconnecting += (error) =>
        {
            ConnectionStateChanged?.Invoke(this, "reconnecting");
            return Task.CompletedTask;
        };

        _hubConnection.Reconnected += (connectionId) =>
        {
            ConnectionStateChanged?.Invoke(this, "connected");
            return Task.CompletedTask;
        };

        _hubConnection.Closed += (error) =>
        {
            ConnectionStateChanged?.Invoke(this, "disconnected");
            return Task.CompletedTask;
        };
    }

    public async Task SendOfferAsync(string targetConnectionId, object offer)
    {
        EnsureHubConnected();
        await _hubConnection!.InvokeAsync("SendOffer", targetConnectionId, offer);
    }

    public async Task SendAnswerAsync(string targetConnectionId, object answer)
    {
        EnsureHubConnected();
        await _hubConnection!.InvokeAsync("SendAnswer", targetConnectionId, answer);
    }

    public async Task SendIceCandidateAsync(string targetConnectionId, object candidate)
    {
        EnsureHubConnected();
        await _hubConnection!.InvokeAsync("SendIceCandidate", targetConnectionId, candidate);
    }

    public async Task UpdateStatusAsync(bool isMuted, bool isVideoEnabled)
    {
        if (string.IsNullOrEmpty(_sessionId)) return;

        EnsureHubConnected();
        await _hubConnection!.InvokeAsync("UpdateStatus", _sessionId, isMuted, isVideoEnabled);
    }

    public async Task SendMessageAsync(string message)
    {
        if (string.IsNullOrEmpty(_sessionId)) return;

        EnsureHubConnected();
        await _hubConnection!.InvokeAsync("SendMessage", _sessionId, message);
    }

    #endregion

    #region Helpers

    private void EnsureAuthenticated()
    {
        if (string.IsNullOrEmpty(_token))
        {
            throw new InvalidOperationException("Not authenticated. Please login first.");
        }
    }

    private void EnsureHubConnected()
    {
        if (_hubConnection?.State != HubConnectionState.Connected)
        {
            throw new InvalidOperationException("SignalR hub is not connected.");
        }
    }

    #endregion

    #region Dispose

    public async ValueTask DisposeAsync()
    {
        if (_hubConnection != null)
        {
            await _hubConnection.StopAsync();
            await _hubConnection.DisposeAsync();
        }

        _httpClient.Dispose();
    }

    #endregion
}
